astroscrappy/astroscrappy.pyx

```diff
@@ -183,14 +209,26 @@ def detect_cosmics(indat, inmask=None, float sigclip=4.5, float sigfrac=0.3,
         mask = np.empty((ny, nx), dtype=np.uint8, order='C')
         mask[:, :] = inmask[:, :]
 
-    # Add back in the previously subtracted sky level and multiply by the gain
-    # The statistics only work properly with electrons.
-    cleanarr += pssl
+    # Multiply by the gain; the statistics only work properly with electrons.
     cleanarr *= gain
 
+    # Copy any noise estimates so we can clean them like the data, otherwise
+    # use the main data array instead.
+    if var is None:
+        if bkg is None:
+            clean_var = cleanarr + readnoise * readnoise
+        else:
+            clean_var = cleanarr - bkg * gain + readnoise * readnoise
+    else:
+        clean_var = var * gain * gain - bkg * gain
```

Comment:
I'm not clear on why you would need to subtract the background from the variance array. If someone has previously calculated the variance in their own custom way, presumably it does not include the background. I see that the docstring has some mention of processing the variance, but I would not have guessed it should include the background.

Reply:
The background is subtracted for the median filtering, which is mostly useful for sky lines. But why should the variance not include the background ? Maybe it depends on what we call "background" ?

Reply:
Sorry, I wrote this in a confusing and possibly wrong way. Let's assume Poisson statistics. The variance of a pixel is the number of electrons in that pixel. If you then subtract some background value from that pixel, the variance of the subtracted pixel does not change; it is still the number of electrons in the unsubtracted pixel. So I don't understand when we would want to use this "background-subtracted variance" array. (Another way to think of it is that the units of this equation don't even make sense: you're subtracting electrons from electrons^2.)

Reply:
The background is subtracted only temporarily, to protect sky lines etc. from being blurred by the median filtering operation, and then gets restored afterwards (L302). I believe the median filtering is necessary to remove cosmic ray flux from the noise estimate; what we want to know to set a sigma threshold is not the actual noise in the pixel but the expected noise in the absence of cosmic rays. However, I do also still have some doubts about the generality of this that I'll elaborate on tomorrow if possible.

Reply:
This is without a doubt the most confusing part of the code. The original concern was that in the imaging algorithm, the variance gets smoothed before being used to calculate the S/N for cosmic rays. This is basically fine in the imaging case. For spectra, if your sky lines are narrow (which they often are), you end up smoothing out the variance from the lines and artificially deflate the variance on sharp sky lines making it more likely to detect them as false positives. What this does is separate the variance into two components, a "known" background component that we assume is poisson and an "unknown" component that might include cosmic rays. We then only smooth the "unknown" variance component. When we calculate the S/N though starting on line 302 we add the unsmoothed variance back from the background. The units look strange just because we assume poisson. 

I would be very happy to make this clearer in the docs/code so happy to have suggestions on that.